runtime/core/decoder/onnx_asr_model.cc

```diff
@@ -0,0 +1,418 @@
+      Ort::MemoryInfo::CreateCpu(OrtDeviceAllocator, OrtMemTypeCPU);
+  if (num_left_chunks_ > 0) {
+    int required_cache_size = chunk_size_ * num_left_chunks_;
+    offset_ = required_cache_size;
+    att_cache_.resize(num_blocks_ * head_ * required_cache_size *
+                          encoder_output_size_ / head_ * 2,
+                      0.0);
+    const int64_t att_cache_shape[] = {num_blocks_, head_, required_cache_size,
+                                       encoder_output_size_ / head_ * 2};
+    att_cache_ort_ = Ort::Value::CreateTensor<float>(
+        memory_info, att_cache_.data(), att_cache_.size(), att_cache_shape, 4);
+  } else {
+    att_cache_.resize(0, 0.0);
+    const int64_t att_cache_shape[] = {num_blocks_, head_, 0,
+                                       encoder_output_size_ / head_ * 2};
+    att_cache_ort_ = Ort::Value::CreateTensor<float>(
+        memory_info, att_cache_.data(), att_cache_.size(), att_cache_shape, 4);
+  }
+
+  // Reset cnn_cache
+  cnn_cache_.resize(
+      num_blocks_ * encoder_output_size_ * (cnn_module_kernel_ - 1), 0.0);
+  const int64_t cnn_cache_shape[] = {num_blocks_, 1, encoder_output_size_,
+                                     cnn_module_kernel_ - 1};
+  cnn_cache_ort_ = Ort::Value::CreateTensor<float>(
+      memory_info, cnn_cache_.data(), cnn_cache_.size(), cnn_cache_shape, 4);
+}
+
+void OnnxAsrModel::ForwardEncoderFunc(
+    const std::vector<std::vector<float>>& chunk_feats,
+    std::vector<std::vector<float>>* out_prob) {
+  Ort::MemoryInfo memory_info =
+      Ort::MemoryInfo::CreateCpu(OrtDeviceAllocator, OrtMemTypeCPU);
+  // 1. Prepare onnx required data, splice cached_feature_ and chunk_feats
+  // chunk
+  int num_frames = cached_feature_.size() + chunk_feats.size();
+  const int feature_dim = chunk_feats[0].size();
+  std::vector<float> feats;
+  for (size_t i = 0; i < cached_feature_.size(); ++i) {
+    for (size_t j = 0; j < feature_dim; ++j) {
+      feats.emplace_back(cached_feature_[i][j]);
+    }
+  }
+  for (size_t i = 0; i < chunk_feats.size(); ++i) {
+    for (size_t j = 0; j < feature_dim; ++j) {
+      feats.emplace_back(chunk_feats[i][j]);
+    }
+  }
+  const int64_t feats_shape[3] = {1, num_frames, feature_dim};
+  Ort::Value feats_ort = Ort::Value::CreateTensor<float>(
+      memory_info, feats.data(), feats.size(), feats_shape, 3);
+  // offset
+  int64_t offset_int64 = static_cast<int64_t>(offset_);
+  Ort::Value offset_ort = Ort::Value::CreateTensor<int64_t>(
+      memory_info, &offset_int64, 1, std::vector<int64_t>{}.data(), 0);
+  // required_cache_size
+  int64_t required_cache_size = chunk_size_ * num_left_chunks_;
+  Ort::Value required_cache_size_ort = Ort::Value::CreateTensor<int64_t>(
+      memory_info, &required_cache_size, 1, std::vector<int64_t>{}.data(), 0);
+  // att_mask
+  Ort::Value att_mask_ort{nullptr};
+  if (num_left_chunks_ > 0) {
+    std::vector<uint8_t> att_mask(required_cache_size + chunk_size_, 1);

```

Comment:
1. move line239 to line237, we need to keep att_mask alive during the lifetime of the current chunk.
2. std::vector\<char\> att_mask(.., ..);